File: July-18-2025/reverseList.js
var numbers = [1, 2, 3, 4]

class ListNode {
  constructor(val) { 
    this.val = val; 
    this.prev = null; 
    this.next = null;
  }
}



function main(numbers) { 
  let head = new ListNode(numbers[0])
  let current = head

  for(let i = 1; i < numbers.length; i++){
    let newNode = new ListNode(numbers[i])
    current.next = newNode;
    current = newNode
  }
  let curr = head;
  while(curr){ 
    console.log(curr.val);
    curr = curr.next
  }
  return head
}


function reverseList(head){ 
  let prev = null; 
  let current = head; 

  while(current){ 
    let nextNode = current.next;
    current.next = prev; 
    prev = current; 
    current = nextNode;
  }
  let curr = prev; 
  while(curr){ 
    console.log(curr.val)
    curr = curr.next
  }
}

let head = (main(numbers))
console.log(reverseList(head))

File: 2025/December/December-19-2025/reverseList.py
from typing import List, Optional


input = [1,2,3,4]

class ListNode: 
  def __init__(self, value): 
    self.value = value
    self.prev = None 
    self.next = None 


def doubleyLinked(input: List[int]) -> Optional[ListNode]: 
  head = ListNode(input[0])
  curr = head

  for i in range(1, len(input)): 
    nextNode = ListNode(input[i])
    curr.next = nextNode
    nextNode.prev = curr
    curr = nextNode
  return head


head = doubleyLinked(input)

curr = head
while curr: 
  print(curr.value, end=' -> ')
  curr = curr.next 
print('NONE')


def reverseList(head: Optional[ListNode]) -> Optional[ListNode]: 
  curr = head
  new_head = None 

  while curr: 
    curr.prev, curr.next = curr.next, curr.prev
    new_head = curr 
    curr = curr.prev
  return new_head


headII = reverseList(head)

curr = headII
while curr: 
  print(curr.value, end=' -> ')
  curr = curr.next
print('NONE')


